File: src/Ch13_StatisticalDescriptionOfData/student_ttest/significantly_different_variances/tutest.py
import numpy as np
import os
import sys


sys.path.append(os.path.realpath('../../avevar'))
from avevar import avevar

sys.path.append(os.path.realpath('../../../Ch6_SpecialFunctions/betai'))
from betai import betai


def tutest(data1, data2):
    # Given the arrays "data1" and "data2", returns the
    # Student's "t" and its significance as "prob".
    # Data are assumed to be drawn from populations with
    # the same true variance.

    # Small values of "prob" indicates that the arrays
    # have different means.
    n1 = len(data1)
    n2 = len(data2)

    ave1, var1 = avevar(data1)
    ave2, var2 = avevar(data2)

    # print(ave1, ave2)
    # print(var1, var2)

    # Student's t
    t = (ave1 - ave2) / np.sqrt(var1/n1 + var2/n2)

    df = (var1/n1 + var2/n2)**2 / ((var1/n1)**2 / (n1 - 1) + (var2/n2)**2 / (n2 - 1))
    # significance
    prob = betai(0.5 * df, 0.5, df/(df + t**2))

    return t, prob

File: src/numericalrecipes/Ch13_StatisticalDescriptionOfData/student_ttest/different_variances/tutest.f90
module mod_tutest
   ! Contains the following subroutine: tutest
   implicit none
   contains

   subroutine tutest(data1, n1, data2, n2, t, prob)
      ! Given the arrays "data1" and "data2" of lengths "n1" and "n2", returns
      ! the Student's "t" and its significance as "prob".
      ! Data are assumed to be drawn from populations with unequal variances.
      ! Small values of "prob" indicates that the arrays have different means.
      use mod_betai
      use mod_avevar
      implicit none
      real(8), intent(in)                  :: data1(n1), data2(n2)
      integer, intent(in)                  :: n1, n2
      real(8)                              :: ave1, var1, ave2, var2, &
                                              df, var, t, prob

      call avevar(data1, n1, ave1, var1)
      call avevar(data2, n2, ave2, var2)

      ! Degrees of freedom
      df = (var1 / n1 + var2 / n2)**2 / ((var1 / n1)**2 / (n1 - 1) + &
                                         (var2 / n2)**2 / (n2 - 1))

      ! Student's t
      t = (ave1 - ave2) / sqrt(var1 / n1 + var2 / n2)

      ! Significance
      prob = betai(0.5d0 * df, 0.5d0, df / (df + t**2))
   end subroutine tutest

end module mod_tutest
